### ISSUE TITLE
ValidateContext should give better suggestions

### ISSUE THREAD

**Issue by jwnimmer-tri** [RobotLocomotion/drake#17265]
When a user passes a `Context` object to a `System` function, we use `ValidateContext` to check that the context actually matches that particular `System` instance.  It's a common error for users to pass a diagram context into the MbP, or vice versa.  We always _detect_ that error (per #11801), but we don't give good suggestions about how to resolve the problem.

The victory here is that the exception message contains a suggestion something like "You passed the Diagram root context into the MbP system; use GetMyContextFromRoot instead."

We should be able to accomplish that by crawling around the diagrams (in both directions) and diagram contexts (in both directions) to figure out what matched up with what, to give a more specific suggestion.

Ditto for `ValidateCreatedForThisSystem`.

\CC @sherm1 FYI

**Comment by SeanCurtis-TRI** [comment]
If you have no complaints, I'm going to use this to keep my pipeline full.

**Comment by SeanCurtis-TRI** [comment]
@jwnimmer-tri I'm looking for elaboration on the initial description.

First an exploration into current events:

If I go with the most classic version of `Context` mis-use (passing the root `Diagram`'s `Context` to the `MultibodyPlant`), I currently get a message that looks like this (line break added for reader's convenience):

```
Context was not created for drake::multibody::MultibodyPlant<double> system ::_::plant;
it was created for system ::_
```

As I contrast this error message with the issue description, what appears to need attention is:

1. Adding the guidance to call `GetMyContextFromRoot()`.
2. The semantics that the context provided belongs to the "root context".

Furthermore, as I apply my own understanding, I'd add the following observations:

3. It doesn't seem we have these diagram "paths" as first-class citizens of Drake concepts. Thus, we get names like` ::_` which are probably not meaningful to most Drake users. Are these names truly helpful?

As I see it: (1) is trivial, (2) requires elaboration -- under what rules would we swap path names for something like "Diagram root", (3) are we simply jettisoning the path names entirely from the message? If so, (3) gets resolved for free.

Finally, you allude "crawling around the diagrams". This is highly suggestive in terms of relationships. By doing so, we could provide text like:
  - the root diagram
  - This system's parent diagram
  - This system's [descendant] system†
  - This system's sibling|cousin†
  - etc.

† For these cases, the "guidance" message could be weirder in this context because the solution cannot be couched in terms of any of the systems or contexts in the scope of the operation.

It seems that walking the diagram casts a wide net, what were you hoping to catch?

(Foot note: the validation of other system quantities (state, parameters, etc.) has a different message that doesn't mention what the quantity is associated with at all. This will require the greater overhaul to match the same kind of relationship-describing behavior of `Context` validation.)

**Comment by jwnimmer-tri** [comment]
In my dream world, we can easily traverse the parent-child graph of Contexts (and/or Systems) to give a really perfect suggestion -- or even calculate "what the user meant to say" automatically at runtime, with no throwing.

For the purpose of this ticket, though, I think we can get 95% of the benefit just by special-casing the root.  If either the System or the Context is the root one (has no parent), then we give a more specific message.  Otherwise, we can keep the same message.

All of the messages should probably hyperlink to a [troubleshooting](https://drake.mit.edu/troubleshooting.html) FAQ entry.

For the root-specific message, I don't see a need to print the `::_` name; as you note, it's usually not set to anything meaningful.

In short, I think focusing on the most common error situations should be sufficient to close this.
